flow/multiagent_envs/traffic_light_grid.py

```diff
@@ -0,0 +1,260 @@
+                       grid_array["inner_length"])
+
+        # TODO(cathywu) refactor PO_TrafficLightGridEnv with convenience
+        # methods for observations, but remember to flatten for single-agent
+
+        # Observed vehicle information
+        speeds = []
+        dist_to_intersec = []
+        edge_number = []
+        all_observed_ids = []
+        for _, edges in self.scenario.node_mapping:
+            local_speeds = []
+            local_dists_to_intersec = []
+            local_edge_numbers = []
+            for edge in edges:
+                observed_ids = \
+                    self.k_closest_to_intersection(edge, self.num_observed)
+                all_observed_ids.append(observed_ids)
+
+                # check which edges we have so we can always pad in the right
+                # positions
+                local_speeds.extend(
+                    [self.k.vehicle.get_speed(veh_id) / max_speed for veh_id in
+                     observed_ids])
+                local_dists_to_intersec.extend([(self.k.scenario.edge_length(
+                    self.k.vehicle.get_edge(
+                        veh_id)) - self.k.vehicle.get_position(
+                    veh_id)) / max_dist for veh_id in observed_ids])
+                local_edge_numbers.extend([self._convert_edge(
+                    self.k.vehicle.get_edge(veh_id)) / (
+                    self.k.scenario.network.num_edges - 1) for veh_id in
+                                           observed_ids])
+
+                if len(observed_ids) < self.num_observed:
+                    diff = self.num_observed - len(observed_ids)
+                    local_speeds.extend([0] * diff)
+                    local_dists_to_intersec.extend([0] * diff)
+                    local_edge_numbers.extend([0] * diff)
+
+            speeds.append(local_speeds)
+            dist_to_intersec.append(local_dists_to_intersec)
+            edge_number.append(local_edge_numbers)
+
+        # Edge information
+        density = []
+        velocity_avg = []
+        for edge in self.k.scenario.get_edge_list():
+            ids = self.k.vehicle.get_ids_by_edge(edge)
+            if len(ids) > 0:
+                # TODO(cathywu) Why is there a 5 here?
+                density += [5 * len(ids) / self.k.scenario.edge_length(edge)]
+                velocity_avg += [np.mean(
+                    [self.k.vehicle.get_speed(veh_id) for veh_id in
+                     ids]) / max_speed]
+            else:
+                density += [0]
+                velocity_avg += [0]
+        density = np.array(density)
+        velocity_avg = np.array(velocity_avg)
+        self.observed_ids = all_observed_ids
+
+        # Traffic light information
+        last_change = self.last_change.flatten()
+        direction = self.direction.flatten()
+        currently_yellow = self.currently_yellow.flatten()
+        # Dummy values for lights that are out of range
+        # TODO(cathywu) are these values reasonable?
```

Comment:
Sorry, what is this supposed to do?

Reply:
Ah, I'll add a comment. This is a catch-all for when the relative_node method returns a -1 (when there is no node in the direction sought). So I add a last item to the lists here, which will serve as a default value.